### DISCUSSION TITLE
Online "REPL"

### DISCUSSION THREAD

**Discussion by bmeck** [souffle-lang/souffle#2359] [Show and tell]
We have been working with Soufflé a bit and came up with a need for a WASM build. I just wanted to throw together a simple example REPL to see how things ended up.

We ended up having to modify the CMake file a bit and disable features to get it to work, but at 1.2MB over the wire it is still quite usable for small workloads even without things like parallel compute and the C Preprocessor, interestingly it can be made to generate C++ still with a bit of tweaks. It lacks highlighting and other niceties due to just being a demo of sorts, but it was still fun to see. I'm not going to go too much farther, but it might be nice to support a WASM build for niche situations.

https://user-images.githubusercontent.com/234659/201383471-f0c85aeb-c39e-4815-8c1c-587a810c693b.mov

**Comment by philzook58** [comment]
I have also experimented with this some. I thought the emscripten builds were still working with no editing.

- https://github.com/souffle-lang/souffle/pull/2297
- http://www.philipzucker.com/souffle-web/
- https://www.philipzucker.com/souffle/
- https://www.philipzucker.com/datalog-book/

**Reply by bmeck** [reply]
I can copy paste the commands in that build PR and it works, but my toying around with it yesterday was having build failures from things like `emcmake` not supporting `--build` while the PR doesn't build using `emcmake`.

**Reply by bmeck** [reply]
Also spent quite a while seeing aborts from lack of c preprocessor that were effectively silent (set process.exitCode=1 then stopped execution, no stderr or anything).

**Reply by philzook58** [reply]
Yes, the c preprocessor is not include in the PR. The example javascript has the flag "--no-preprocessor" to avoid it, but if your example uses CPP it won't work unless you preprocess it yourself in browser with some javascript CPP replacement first

**Reply by philzook58** [reply]
My impression is that one generally uses emcmake for only the initial build directory setup command and after the build directory is setup you use cmake without emcmake

**Comment by b-scholz** [comment]
We would need a new demo server urgently. Unfortunately, our old uni server is defunct. If we could find volunteers for this project, that would be very welcome.

**Reply by bmeck** [reply]
@b-scholz these both run on regular html so no need for a special server. Docs site could host even if completely static.

**Comment by b-scholz** [comment]
Could you help rewriting this page

https://souffle-lang.github.io/simple

 so that it works in the browser?

**Reply by bmeck** [reply]
@b-scholz wouldn't be too hard, but looking at the doc site it would need to be able to pull in a WASM build of soufflé doing so would be possible if it had a git submodule of this repo and was built in a github action most likely using same commands as @philzook58 pointed to above. Hooking up the HTML/JS glue wouldn't be too hard after that.

**Reply by bmeck** [reply]
not too bad to hook up but some things are not well suited for the browser/IDE and would require changes to the examples to be more ergonomic (like tab based delimiters by default). Not going to make a build script / github action but can get scaffold and example of what to migrate examples to setup sometime. unclear exact UX to use for stdout/stderr here 🤷🏼 .

https://user-images.githubusercontent.com/234659/204662292-38b3c807-4f2f-47ee-8c0a-186f9392bad3.mov

**Comment by b-scholz** [comment]
It might be good to build the WASM executable for each release as GitHub actions and deploy it on souffle-lang.github.io.
